tools/ncf_constraints.py

```diff
@@ -0,0 +1,90 @@
+import re
+import subprocess
+import json
+import os.path
+import shutil
+import sys
+import os
+import codecs
+
+
+### Constraint checking function 
+
+def from_list( parameter, accepted_result):
+  return parameter in accepted_result
+
+def max_length( parameter, max_size):
+  return len(parameter) <= max_size
+
+def min_length( parameter, min_size):
+  return len(parameter) >= min_size
+
+def match_regex(parameter, regex):
+  match = re.search(regex, parameter)
+  return match is not None
+
+def not_match_regex(parameter, regex):
+  return not match_regex(parameter, regex)
+
+def check_parameter(parameter, constraint):
+  """Checks that a parameter is ok with the constraint passed as parameter"""
+
+  result = True
+  errors = [];
+
+  # First check 'string' constraints
+  if not constraint.get("allow_empty_string", False):
+    if match_regex(parameter, "^$"):
+      errors.append("allow_empty_string")
+      result = False
+
+  if not constraint.get("allow_whitespace_string", False):
+    if match_regex(parameter, r'^\s') or match_regex(parameter, r'\s+$'):
+      errors.append("allow_whitespace_string")
+      result = False
+
+  # Check if in list of accepted values
+  if "list" in constraint:
+    if not from_list(parameter, constraint["list"]):
+      errors.append("list")
+      result = False
+
+  # Check regexs constraints
+  if "regex" in constraint:
+    if not match_regex(parameter, constraint["regex"]):
+      errors.append("regex")
+      result = False
+
+  if "not_regex" in constraint:
+    if not not_match_regex(parameter, constraint["not_regex"]):
+      errors.append("not_regex")
+      result = False
+
+  # Check length constraints
+  if "min_length" in constraint:
+    if not min_length(parameter, constraint["min_length"]):
+      errors.append("min_length")
+      result = False
+
+  if "max_length" in constraint:
+    if not max_length(parameter, constraint["max_length"]):
+      errors.append("max_length")
+      result = False
+  
```

Comment:
peckpeck wrote:
> There have been not release with those specs, so they can still change.

Whoa! Well, yes, of course we can change them. I'm not trying to start a war here, just trying to explain to you how we got to this implementation.

I don't know why you weren't with us when we discussed the specs, but a summary of them was shared by email weeks ago (April 15th, to be precise). I would have expected to have this discussion at the time. This is a *technical* review, so rather off topic to discuss specs (as a reminder, the goal of a TR is to make sure the implementation works and matches the specs).

However, I am happy to rediscuss the specs if you feel the naming or logic is flawed. As usual, I'd much rather spend more time to make something good than have to support something bad for years.

My understanding is that you're querying two points here:

1) Naming: the main reason for this PR is to allow to trim leading/trailing whitespace and/or whitespace only values. All values are currently automatically trimmed by ncf-builder, and we want to be able to tell ncf-builder to not do this via this feature. I agree that "allowing" things isn't really a *constraint*, more a "relaxation of an implicit constraint". However, we failed to find a better name. Do you have any ideas?

2) The need to have an extra "or" operator as well as the implicit "and" operator.

Did I understand your comments correctly?

One answer in particular:
> Adding other constraints with a logical AND to a list is not really useful, [...] a list is already a full set of constraint in itself and adding some can only confirm existing constraints or contradict them.

I agree, it's not particularly useful to have "list" and other constraints. But I don't see that as a problem - if users don't want to use both, they won't. However, each taken separately *is* useful, therefore I think that makes this a valid feature.